runtime/syntax/cpp.yaml

```diff
@@ -47,9 +47,9 @@ rules:
     - constant.string:
         start: "'"
         end: "'"
-        skip: "(\\\\.)|(.*[^']$)"
+        skip: "(\\\\.)|([^']{4}'[^']{4})"
```

Comment:
...And this doesn't make sense to me at all, and doesn't work at all. Even a very basic case like `123'45` is not highlighted as a number.

Instead of these hacks, it seems that the `skip` regex should contain the actual regex for a number containing single quotes (to ensure that it, well, actually skips all such numbers). I've tried this:

```
        skip: "(\\\\.)|(\\b[1-9][0-9]*'[0-9']*[0-9]\\b)"
```

and it seems to work fine, although only for decimal integers without `U` and `L` suffixes. So also need to add the rest here: hexadecimals, floating point, `U` and `L` suffixes and so on. Yes, the resulting regex is gonna look horrifying, but it seems there is no other way to make it actually work.

Reply:
You are correct that `123'45` won't work. But your change will not work as well since the char highlighting will "leak" to the next line.

![image](https://github.com/zyedidia/micro/assets/93885501/43796edb-21e3-4201-bd40-ce1f257b0126)

`testBinary` and `testHex` if statement is leaking in this case. 
The main point of my change is to make it not "leak" and it's quite tricky to get right.

Reply:
Here's the test case I am using atm.

```c++
char testGoodChar = 'a'
char testBadChar = 'aba'
char testGoodInvis = '\t'
char testBadInvis = '\tn'
int testBinary = 0b0000'0000;
int testHex = 0x0AB0'aA20;
int testNormal = 2;

//Good char if statement
if(testChar[i] == '.')

//Good invisible char if statement
if(testChar[i] == '\t')

//Bad invisible char if statement
if(testChar[i] == '\tn' && 'fs')

const std::unordered_set<char> CharTokens = {'(', ')', ';', '{', '}', '/', '#', ',', '=', '<', '>'};

int test = 123'45;

//Bad char if statement
if(testChar[i] == '.sf')

//Good hex If statement
if(testHex[i] == 0x1234'1234)

//Good binary If statement
if(testBin[i] == 0b0000'1111)
```

Reply:
I'm not sure what exactly do you mean by "leaking", but I already see that my change doesn't work correctly in such a basic case:

```
123'45 'j'
```

`'j'` is highlighted as error, while it is obviously a valid character literal.

But, I've tried wit @JoeKar 's PR #3127 and my change works correctly with it.

So it is yet another misbehavior of micro's highlighting of syntax regions, which should be fixed by #3127 once we merge it (after I find some time to properly review it, which I'm constantly struggling to).

Reply:
Take a look at the screenshot I replied with, line -21 and the line my cursor is on (the one with 26) is highlighting characters to the next line until it meets the next `'`. Yellow text is the character highlighting.

This is the main thing I want to address in PR mainly since working on a codebase with `'` as a separator for binary or hex is very inconvenient because it will highlight anything until it hits the next `'`

Reply:
So, this is how your example looks with the newest version of your PR:

![image](https://github.com/zyedidia/micro/assets/147101/70291cd7-198e-449c-8efb-1ca6d5afe261)

And this is how it looks with #3127 + the following change:

```
--- a/runtime/syntax/cpp.yaml
+++ b/runtime/syntax/cpp.yaml
@@ -30,7 +30,7 @@ rules:
       # Parenthetical Color
     - symbol.brackets: "[(){}]|\\[|\\]"
       # Integer Literals
-    - constant.number: "(\\b([1-9][0-9']*|0[0-7']*|0[Xx][0-9a-fA-F']+|0[Bb][01]+)([Uu]?[Ll][Ll]?|[Ll][Ll]?[Uu]?)?\\b)"
+    - constant.number: "(\\b([1-9][0-9']*|0[0-7']*|0[Xx][0-9a-fA-F']+|0[Bb][01']+)([Uu]?[Ll][Ll]?|[Ll][Ll]?[Uu]?)?\\b)"
       # Decimal Floating-point Literals
     - constant.number: "(\\b(([0-9']*[.][0-9']+|[0-9']+[.][0-9']*)([Ee][+-]?[0-9']+)?|[0-9']+[Ee][+-]?[0-9']+)[FfLl]?\\b)"
       # Hexadecimal Floating-point Literals
@@ -47,7 +47,7 @@ rules:
     - constant.string:
         start: "'"
         end: "'"
-        skip: "\\\\."
+        skip: "(\\\\.)|(\\b[1-9][0-9]*'[0-9']*[0-9]|0[0-7]+'[0-7']*[0-7]|0[Xx][0-9A-Fa-f]+'[0-9A-Fa-f']*[0-9A-Fa-f]|0[Bb][01]+'[01']*[01]\\b)"
         rules:
             - error: "..+"
             - constant.specialChar: "\\\\([\"'abfnrtv\\\\]|[0-3]?[0-7]{1,2}|x[0-9A-Fa-f]{1,2}|u[0-9A-Fa-f]{4}|U[0-9A-Fa-f]{8})"
```

![image](https://github.com/zyedidia/micro/assets/147101/a455ad56-c0a8-48d1-9ef3-b5db690a7a97)

Reply:
Thanks, that indeed works :+1: 
I have modified the changes to that for skip

Reply:
> [...] after I find some time to properly review it [...]

And I fear that moment, when you finish the review. :smiling_face_with_tear: